cynthion/python/examples/tutorials/gateware-usb-device-04.py

```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env python3
+#
+# This file is part of Cynthion.
+#
+# Copyright (c) 2024 Great Scott Gadgets <info@greatscottgadgets.com>
+# SPDX-License-Identifier: BSD-3-Clause
+
+from amaranth                                    import *
+from amaranth.lib.fifo                           import SyncFIFO
+from luna.usb2                                   import USBDevice
+from usb_protocol.emitters                       import DeviceDescriptorCollection
+
+from luna.gateware.usb.request.windows           import (
+    MicrosoftOS10DescriptorCollection,
+    MicrosoftOS10RequestHandler,
+)
+from usb_protocol.emitters.descriptors.standard  import get_string_descriptor
+from usb_protocol.types.descriptors.microsoft10  import RegistryTypes
+
+from luna.gateware.stream.generator              import StreamSerializer
+from luna.gateware.usb.request.control           import ControlRequestHandler
+from luna.gateware.usb.request.interface         import SetupPacket
+from luna.gateware.usb.usb2.request              import RequestHandlerInterface
+from luna.gateware.usb.usb2.transfer             import USBInStreamInterface
+from usb_protocol.types                          import USBRequestType
+
+from luna.usb2                                   import USBStreamInEndpoint, USBStreamOutEndpoint
+from usb_protocol.types                          import USBDirection, USBTransferType
+
+VENDOR_ID  = 0x1209 # https://pid.codes/1209/
+PRODUCT_ID = 0x0001
+
+MAX_PACKET_SIZE = 512
+
+class VendorRequestHandler(ControlRequestHandler):
+    VENDOR_SET_FPGA_LEDS   = 0x01
+    VENDOR_GET_USER_BUTTON = 0x02
+
+    def elaborate(self, platform):
+        m = Module()
+
+        # shortcuts
+        interface: RequestHandlerInterface = self.interface
+        setup: SetupPacket = self.interface.setup
+
+        # get a reference to the FPGA LEDs and USER button
+        fpga_leds   = Cat(platform.request("led", i).o for i in range(6))
+        user_button = platform.request("button_user").i
+
+        # create a streamserializer for transmitting IN data back to the host
+        serializer = StreamSerializer(
+            domain           = "usb",
+            stream_type      = USBInStreamInterface,
+            data_length      = 1,
+            max_length_width = 1,
+        )
+        m.submodules += serializer
+
+        # we've received a setup packet containing a vendor request.
+        with m.If(setup.type == USBRequestType.VENDOR):
+            # take ownership of the interface
+            m.d.comb += interface.claim.eq(1)
```

Comment:
This can cause conflicts with the Microsoft handler:

https://github.com/greatscottgadgets/luna/blob/1ad9967413bfd3e1047711ae48b4d610736c5e05/luna/gateware/usb/request/windows/__init__.py#L56-L62

We could only assert `claim` when `request` is `VENDOR_SET_FPGA_LEDS` or `VENDOR_GET_USER_BUTTON`. There's also no need for the `UNHANDLED` case.

Reply:
Nice, tx @mndza!

Fixed in: cecd5cf